hedera-node/src/main/java/com/hedera/services/ledger/BalanceChange.java

```diff
@@ -0,0 +1,184 @@
+ * You may obtain a copy of the License at
+ * 
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ * 
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * ‍
+ */
+
+import com.google.common.base.MoreObjects;
+import com.hedera.services.store.models.Id;
+import com.hedera.services.store.models.NftId;
+import com.hederahashgraph.api.proto.java.AccountAmount;
+import com.hederahashgraph.api.proto.java.AccountID;
+import com.hederahashgraph.api.proto.java.NftTransfer;
+import com.hederahashgraph.api.proto.java.ResponseCodeEnum;
+import com.hederahashgraph.api.proto.java.TokenID;
+import org.apache.commons.lang3.builder.EqualsBuilder;
+import org.apache.commons.lang3.builder.HashCodeBuilder;
+
+import static com.hederahashgraph.api.proto.java.ResponseCodeEnum.INSUFFICIENT_ACCOUNT_BALANCE;
+import static com.hederahashgraph.api.proto.java.ResponseCodeEnum.INSUFFICIENT_TOKEN_BALANCE;
+import static com.hederahashgraph.api.proto.java.ResponseCodeEnum.SENDER_DOES_NOT_OWN_NFT_SERIAL_NO;
+
+/**
+ * Process object that encapsulates a balance change, either ℏ or token unit .
+ *
+ * Includes an optional override for the {@link ResponseCodeEnum} to be used
+ * in the case that the change is determined to result to an insufficient balance;
+ * and a field to contain the new balance that will result from the change.
+ * (This field is helpful to simplify work done in {@link HederaLedger}.)
+ *
+ * The {@code explicitTokenId} and {@code explicitAccountId} fields are
+ * temporary, needed to interact with the {@link com.hedera.services.ledger.accounts.BackingAccounts}
+ * and {@link com.hedera.services.ledger.accounts.BackingTokenRels} components
+ * whose APIs still use gRPC types.
+ */
+public class BalanceChange {
+	static final TokenID NO_TOKEN_FOR_HBAR_ADJUST = TokenID.getDefaultInstance();
+
+	private final Id token;
+	private final Id account;
+	private final long units;
+	private ResponseCodeEnum codeForInsufficientBalance;
+
+	private long newBalance;
+	private NftId nftId = null;
+	private TokenID tokenId = null;
+	private AccountID accountId;
+	private AccountID counterPartyAccountId = null;
+
+	private BalanceChange(Id token, AccountAmount aa, ResponseCodeEnum code) {
+		this.token = token;
+
+		this.accountId = aa.getAccountID();
+		this.account = Id.fromGrpcAccount(accountId);
+		this.units = aa.getAmount();
+
+		this.codeForInsufficientBalance = code;
+	}
+
+	private BalanceChange(Id token, AccountID sender, AccountID receiver, long serialNo, ResponseCodeEnum code) {
+		this.token = token;
+
+		this.accountId = sender;
+		this.counterPartyAccountId = receiver;
+		this.account = Id.fromGrpcAccount(accountId);
+		this.units = serialNo;
```

Comment:
I think that using `units` for the serial number might introduce confusion. Wouldn't it be better to skip the setting of serialNO to the units property and use the `nftId` directly? For `serialNo()` we could use `nftId.serialNo()`

Reply:
You're definitely right...this class is a mess, and will get even messier with a pending merge in `master`.

I will clean it up carefully after HIP-18 changes are all in `master`. 👍